libs/qCC_io/ShpFilter.cpp

```diff
@@ -1624,69 +1545,51 @@ CC_FILE_ERROR ShpFilter::loadFile(const QString& filename, ccHObject& container,
 	}
 
 	//load shapes
-	CC_FILE_ERROR error = CC_FERR_NO_ERROR;
+	error = CC_FERR_NO_ERROR;
 	ccPointCloud* singlePoints = nullptr;
-	qint64 pos = file.pos();
 	//we also keep track of the polylines 'record number' (if any)
 	QMap<ccPolyline*, int32_t> polyIDs;
 	int32_t maxPolyID = 0;
 	int32_t maxPointID = 0;
 	bool is3DShape = false;
-	while (fileLength >= 12)
+	while (hdr.fileLength - shpStream.device()->pos() > 0)
 	{
-		file.seek(pos);
-		assert(pos + fileLength == fileSize);
-		//load shape record in main SHP file
+		if (shpStream.status() != QDataStream::Ok )
 		{
-			file.read(header, 8);
-			//Byte 0: Record Number
-			int32_t recordNumber = qFromBigEndian<int32_t>(*reinterpret_cast<const int32_t*>(header)); //Record numbers begin at 1
-			//Byte 4: Content Length
-			int32_t recordSize = qFromBigEndian<int32_t>(*reinterpret_cast<const int32_t*>(header + 4)); //Record numbers begin at 1
-			recordSize *= 2; //recordSize is measured in 16-bit words
-			fileLength -= 8;
-			pos += 8;
+			ccLog::Warning("[SHP] Something went wrong reading the file");
+			return CC_FERR_READING;
+		}
+		int32_t recordNumber, recordSize, shapeTypeInt;
+		shpStream.setByteOrder(QDataStream::BigEndian);
```

Comment:
Does this format really switch between "endianness"?

Reply:
As strange as it may seem, yes the endianness switches throughout the file parts

Reply:
Yuck.

That seems like an odd design decision :-)